tt_metal/impl/dispatch/command_queue.hpp

```diff
@@ -466,59 +464,73 @@ class HWCommandQueue {
     friend void EnqueueTrace(Trace& trace, bool blocking);
     friend void EndTrace(Trace& trace);
     friend Trace BeginTrace(CommandQueue& cq);
-    friend void EnqueueProgram(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace);
-    friend void EnqueueReadBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking);
-    friend void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer,
-                                          const void* src, bool blocking);
-    friend void Finish(CommandQueue & cq);
+    friend void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace);
+    friend void EnqueueReadBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking);
+    friend void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, const void* src, bool blocking);
+    friend void FinishImpl(CommandQueue & cq);
     friend class Trace;
 };
 
-class CommandQueue
-{
-    public:
-        CommandQueue () = delete;
-        CommandQueue ( Device * device, uint32_t id) : device_ptr(device), cq_id(id){}
-        ~CommandQueue() {}
-
-        template < class F >
-        tf::AsyncTask& submit ( F && func, std::reference_wrapper< std::shared_future< void > > event ){
-            std::lock_guard<std::mutex> lk(this->mutex);
-            std::tie(this->last, event.get()) = this->last.empty() ? detail::GetExecutor().dependent_async ( std::forward<F>(func) ) : detail::GetExecutor().dependent_async ( std::forward<F>(func), this->last);
-            return this->last;
-        }
 
-        //WARNING: Exceptions thrown by sub-tasks are retieved via future; the silent versions don't return a future object, and therefore exceptions are lost
-        template < class F >
-        void submit ( F && func){
-            std::lock_guard<std::mutex> lk(this->mutex);
-            this->last = this->last.empty() ? detail::GetExecutor().silent_dependent_async(std::forward<F>(func) ) : detail::GetExecutor().silent_dependent_async( std::forward<F>(func), this->last );
-        }
+// Common interface for all command queue types
+struct CommandQueueInterface {
+    EnqueueCommandType type;
+    std::optional<bool> blocking;
+    std::optional<std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>>> buffer;
```

Comment:
The fields are optional because host APIs do not have a consistent interface, so a common interface into the CQ needs to support the superset.

I opted for std::optional for its memory efficiency, on allocation each std::optional should roughly only have fixed and minimal book-keeping entries (has_value, value pointer = null). That makes pushing a command to the CQ fast.

I could see an argument for making blocking required on the interface, so it's made explicit what the blocking/non-blocking behaviour of each command is - even if user doesn't specify it